add new api

zhuyuping opened the pull request:
完成1.1功能开发

Commit:
add new api

Commit:
add scoreLevel
add scoreOffset

plutoless commented on AgoraKaraokeScore/AgoraLrcScoreView/NewApi/Events.swift:
```diff
@@ -0,0 +1,38 @@
+}
+
+/// 分数计算协议
+@objc public protocol ScoreCalculatorProtocol {
```
IScoreAlgorithm

zhuyuping replied:
done

plutoless commented on AgoraKaraokeScore/AgoraLrcScoreView/NewApi/Events.swift:
```diff
@@ -0,0 +1,38 @@
+    /// - Parameters:
+    ///   - models: 字得分信息集合
+    /// - Returns: 计算后的分数 [0, 100]
+    @objc optional func getLineScore(with toneScores: [ToneScoreModel]) -> Int
```
不需要optional，这个方法是必须实现的

zhuyuping replied:
done

plutoless commented on AgoraKaraokeScore/AgoraLrcScoreView/NewApi/KaraokeView.swift:
```diff
@@ -0,0 +1,72 @@
+    /// 设置自定义分数计算对象
+    /// - Note: 如果不调用此方法，则内部使用默认计分规则
+    /// - Parameter calculator: 遵循`ScoreCalculatorProtocol`协议实现的对象
+    public func set(calculator: ScoreCalculatorProtocol) {}
```
没必要用set函数重载，下面一样，这个应该是algorithm

plutoless replied:
如果是score相关的方法，就建议用setScoreXXX，如果是lyric相关就setLyricXXX

plutoless commented on AgoraKaraokeScore/AgoraLrcScoreView/NewApi/KaraokeView.swift:
```diff
@@ -0,0 +1,72 @@
+    /// 设置自定义的等级分布
+    /// - Note: 如果设置空数组，则采用内部默认代替
+    /// - rankConfigItems: 数组，分数从低往高排序。即：[("c", 0..<60>), ("b", 60..<70), ....]
+    public func set(rankConfigItems: [ScoreRankConfigItem]) {}
```
等级分布建议走协议函数，不要这样，这样指定区间会非常麻烦，最好给一个值然后直接返回一个level

plutoless replied:
可以直接作为IScoreAlgorithm的一部分

plutoless commented on AgoraKaraokeScore/AgoraLrcScoreView/NewApi/KaraokeView.swift:
```diff
@@ -0,0 +1,72 @@
+    /// 设置自定义的激励分布
+    /// - Note: 如果设置空数组，则采用内部默认代替
+    /// - incentiveConfigItems: 数组，分数从低往高排序。即：[("good", 0..<60>), ("vary good", 60..<70), ....]
+    public func set(incentiveConfigItems: [IncentiveConfigItem]) {}
```
类似，不要这样设置

plutoless commented on AgoraKaraokeScore/AgoraLrcScoreView/NewApi/KaraokeView.swift:
```diff
@@ -0,0 +1,72 @@
+    /// 设置打分分值补偿
+    /// - Note: 在计算分值的时候作为补偿
+    /// - Parameter scoreOffset: 分值补偿 [-100, 100], 如不设置默认为0
+    public func set(scoreOffset: Int) {}
```
如果是补偿，则不应该叫offset，这个值得含义具体值什么?

zhuyuping replied:
算法给到的解释是打分分值补偿。具体是算晚分数之后，会加上一个值，这个值由offset作为分子提供：offset/100 * standardPitch

zhuyuping replied:
可用：Compensation

zhuyuping replied:
CompensationOffset

plutoless commented on AgoraKaraokeScore/AgoraLrcScoreView/NewApi/LyricsView.swift:
```diff
@@ -0,0 +1,43 @@
+    /// 歌词上下间距
+    public var lyricLineSpacing: CGFloat = 10
+    /// 等待开始圆点背景色
+    public var waitingViewBackgroundColor: UIColor? = .gray
```
最好能把waitingViewStyle做成一个struct属性，然后细节的属性放进去，这样会干净些

zhuyuping replied:
用一个WaitingInfo封装

zhuyuping replied:
再想一下。待会再给确认

plutoless commented on AgoraKaraokeScore/AgoraLrcScoreView/NewApi/Model.swift:
```diff
@@ -0,0 +1,143 @@
+    case slow
+}
+ 
+public class LyricsModel: NSObject {
```
LyricModel

zhuyuping replied:
done

Commit:
根据评论进行更新

Commit:
更新注释

Commit:
feat：Add Parser
feat: Implement of KaraokeView parser method

Commit:
feat: Add Tests for Parser
feat: Add class FirstToneHintView
feat: Update Demo

Commit:
feat: Add Lyrics view

Commit:
feat: Add drag
feat: Fix Lyrics issue
feat: Add reset

Commit:
feat: maxWidth implement
feat: lyricLineSpacing implement

Commit:
change api

Commit:
feat: Add ScoringCanvasView

implement scoring view

Commit:
feat: Add scoring

Commit:
temp save

Commit:
feat: Add ScoringVM

zhuyuping committed:
Update Parser in scoring

- feat: Add fix time issue(xml file)
- test: Add time issue case

zhuyuping committed:
Add line score calculate

zhuyuping committed:
feat: Add Grade view

zhuyuping committed:
Update `scoreAlgorithm`

zhuyuping committed:
Update calculated line score

zhuyuping committed:
Update Api

- feat: remove api (Grade\Incentive)
- feat: implement of incentiveView
- feat: implement of gradeView

zhuyuping committed:
Update IncentiveView

- feat: Update combo text
- feat: Update LyricView normal and highlight font
- feat: Add Console View, update scoringView/scoringVM delegate
- feat: Update filter count of pitch == 0

zhuyuping committed:
Update KaraokeView

- feat: Update filter of pitch = 0, count = 5
- feat: Add Double extension

zhuyuping committed:
Update debugView

zhuyuping committed:
Update demo

zhuyuping committed:
Update cursor custom api

- feat: Add api about api

zhuyuping committed:
Update Emitter

- fix: fix no cel in emitter, while `cell.lifetime` = 1 after a long time
- feat: Update default color of `textHighlightedColor `
- feat: Add emitterImages
- fix: fix start position of `scoreLabel` in  `localPitchView`

update demo

zhuyuping committed:
Add calculated score after drag

zhuyuping committed:
Fix end of  last line

zhuyuping committed:
Fix Drag issue

zhuyuping committed:
Update LyricsView.swift

zhuyuping committed:
Update demo

zhuyuping committed:
Update Scoring

- feat: Remove 250ms of progress
- feat: Add `VoicePitchChanger`
- feat: Update `GradeView`

- demo: add oc test

zhuyuping committed:
Add Objc

zhuyuping committed:
Update LyricView

- feat: Add `didset` action
- feat: Update `FirstToneHintViewStyle`

zhuyuping committed:
Update property

- feat: Add `func setPitch(pitch: Double, progress: Int)`
- feat: Update hidden of `FirstToneHintView`
- feat: Update `textHighlightedColor `
- feat: Add `reset()` in `LocalPitchView`
- feat: Update `scoreLevel` and `scoreCompensationOffset`
- feat: Update debug view

Update demo

zhuyuping committed:
Update Emitter

- fix: Fix life time timeout issue
- fix: Fix highlightInfos not update
- fix: Fix Cursor not rest, while pitch  = 0

update scoring debugView

zhuyuping committed:
Update log

zhuyuping committed:
Update ViewHeight

zhuyuping committed:
Add Objc

zhuyuping committed:
Update log

zhuyuping committed:
Add setLog

zhuyuping committed:
Upate debug info

zhuyuping committed:
Update calculatedY

- feat: Update calculatedY
- feat: Update log

zhuyuping committed:
Update IncentiveView

zhuyuping committed:
Update LocalPitchView.swift

zhuyuping committed:
Update IncentiveView.swift

zhuyuping committed:
Update GradeView.swift

zhuyuping committed:
Update IncentiveView.swift

zhuyuping committed:
Update log

zhuyuping committed:
Update VoicePitchChanger.swift

zhuyuping committed:
Update demo

zhuyuping committed:
Update code style

zhuyuping committed:
Create GCDTimer.swift

zhuyuping committed:
Update IncentiveView.swift

zhuyuping committed:
Update IncentiveVC.swift

zhuyuping committed:
Update demo

zhuyuping committed:
Update AgoraLyricsScore.podspec

zhuyuping committed:
Update project.pbxproj

plutoless commented on AgoraLyricsScore/Class/KaraokeView.swift:
```diff
@@ -0,0 +1,272 @@
+    ///   - printToConsole: 是否打印到控制台 默认：false
+    ///   - writeToFile: 是否记录在文件 默认：true
+    @objc static public func setLog(printToConsole: Bool, writeToFile: Bool) {
+        Log.setup(printToConsole: printToConsole, writeToFile: writeToFile)
```
如果我想都写呢? 建议用int bit enum

zhuyuping committed:
Update GradeView.swift

- feat: Update colors
- feat: Update C level score

zhuyuping committed:
Update GradeView.swift

zhuyuping committed:
Fix #364 #365 #366

zhuyuping committed:
Add Test

zhuyuping committed:
Fix 367

zhuyuping committed:
Update demo

zhuyuping committed:
Update VoicePitchChanger.swift

zhuyuping committed:
Update ScoreAlgorithm.swift

zhuyuping committed:
Update Hit

- feat: remove pitchDuration
- feat: update score

zhuyuping committed:
Add Unit test

zhuyuping committed:
Update VoicePitchChanger.swift

zhuyuping committed:
Create VoicePitchChangerTests.swift

zhuyuping committed:
Update ViewController.swift

zhuyuping committed:
Update README.md

zhuyuping committed:
Update README.md

zhuyuping committed:
Update README.md

zhuyuping committed:
Update demo

zhuyuping committed:
Update .gitignore

zhuyuping committed:
Update README.md

zhuyuping committed:
Update .gitignore

zhuyuping committed:
Update project.pbxproj

zhuyuping committed:
Update Test

zhuyuping committed:
Fix bundle issue

zhuyuping committed:
Update demo

- feat: Add  app icon
- feat: Update info.plist
- feat: Update rtm token builder

zhuyuping committed:
Fix Lyrics progressRate calculated issue

- fix: Lyrics progressRate calculated issue
- feat: Add Unit Test

zhuyuping committed:
Update MainTestVC.swift

zhuyuping committed:
update method

zhuyuping committed:
Update TestMockScoring.swift

zhuyuping committed:
Update MainTestVC.swift

zhuyuping committed:
Fix LyricView update issue

zhuyuping committed:
Fix：IncentiveView

will show a incentive image ,while score less than 60.
by ENT-436

zhuyuping committed:
Fix bug, one of #456

- fix: drag, after reset, in lyrics view
- feat: update log

update test unit
